### ISSUE TITLE
Allow rendering to offscreen surface without having to create a window

### ISSUE THREAD

**Issue by andrekoehler** [urho3d/urho3d#2189]
I use UrhoSharp and would like to use WPF to draw my UI because the layout system of Urho UI elements lacks a lot of features and is a bit buggy/hard to work with and the Urho editor is also hard to use for UI design. WPF can embed Windows Forms Controls that are used as Urho render target but WPF can can not draw on top of WindowsFormsHost. I would like to render the Urho scene into a bitmap or Direct3D texture and display that with WPF so that I can freely combine it with other WPF content. Unfortunately, Urho requires to either create a new window or use an existing window handle even if I only want to render to offscreen surfaces. I tried using -borderless mode but then the graphics subsystems are disabled and can not be used.

I found the following post which claims that Direct3D11 can render without creating a window, but Direct3D9 can not: [StackExchange post](https://gamedev.stackexchange.com/questions/51526/directx11-without-window-attached)
Not sure if OpenGL can do that too.

It would also be nice if Urho could support rendering without a window for use in server applications.

**Comment by eugeneko** [comment]
Allowing Urho to stay with properly working subsystems and without main window initialized is something like very big kludge across codebase.
Moreover, it won't be generic solution because DX9 doesn't support it.
I doubt that SDL would be happy with it too.
Could you just create invisible window and pass it as holder for Urho main frame?

About the subject: copying bitmaps is ok for rarely updated frames, e.g. previews or so on, but you won't get good FPS if you want to update window in realtime.

**Comment by JSandusky** [comment]
Having been down this road (minus the UrhoSharp part), here are your options:

- **A)** Use HWNDHost and use an IPC scheme to slave everything together
    - Requires some tweaks in Input if you intend to use Urho3D's inputs
        - Can't recall exactly what it was, just a single IF that needed to be tweaked to support slaving the HWND
    - Doesn't work if you need multiple render views
        - SDL2's multiple window support isn't very usable anyways, NVM that it's not supported in Urho3D
- **B)** For DX9 add a new engine construction path intended for use with D3DImage, the buffer provided by D3DImage is now your ruler
    - Use dummy HWND as per the MS docs
        - It's all more or less just following the MS docs for WPF and native DX interop.
    - Have to ditch SDL, use WPF's events and handlers instead
        - Necessary to deal with captures and WPF GUI state being meaningful
        - If you need to run something as though it were in freestanding Urho (like *play* in editor) this means heavily tweaking the input system to accept foreign entries, or gutting the SDL dependence entirely
            - I didn't have to deal with this
- **C)** For DX11 gut [WPFInterop](https://gitlab.com/MarcStan/MonoGame.Framework.WpfInterop/tree/master/MonoGame.Framework.WpfInterop) as it's low level elements are all spot-on
    - Requires new engine construction path for an existing DX device
    - Mostly the same process as **B** with a lot more tedium
    - SharpDX and C++ DX interop is very brittle, SharpDX will be unaware of whatever you've done that invalidates its' C#-side image of what it believes the DX state-machine to be
        - So use as little as possible on the SharpDX side
- Accept the flicker if using DX11 in WPF
     - You can sleep the main thread for a few milliseconds to reduce it but there's no way to force WPF to wait for your DX code to finish pending GPU work, resize and you'll flicker for not being able to paint quickly enough

**A** is the worst, and something you'd only use if getting stuff working right now for a single-view program is more important than anything else.

**B** just sucks because of DX9 and was pretty easy to setup. It also easily turned into **C**.

I went with option **C**, although I had done options **A** and **B** before. Eventually, I just dropped it for MonoGame as I was tripping over Urho3D making things more work than necessary and stubborn to change.

> I found the following post which claims that Direct3D11 can render without creating a window, but Direct3D9 can not: StackExchange post

Dummy HWNDs (0,0,0,0,NULL,NULL,NULL) miss out on most of the message loop and have to hook to get most everything unless recreated. They're the DX9 equivalent of *no-window* as they do little to nothing, they're required on DX9 in order to find a valid device head via the HDC every window gets (the only way to get an *almost certainly valid* one before Vista).

**Comment by pengfei666** [comment]
I have a solution for QT.  Draw Urho3D scene to a framebuffer, then mix the framebuffer content with QML. Then I can make use of all the awesome graphical effects of QML.
This is my repro https://github.com/pengfei666/Urho3D.  See whether it helps you.

**Comment by scaledteam** [comment]
Eng: I am making progress in this direction. I ported the Offscreen driver from SDL, but it only works with EGL. To enable proper EGL support, I updated GLEW to the newest version and forced EGL in CMake, but because of this, it now works only under Wayland and in Offscreen mode. Launching in X11 leads to shader errors. But it works well in my VTuber software in combination with [obs-vkcapture](https://github.com/nowrep/obs-vkcapture/). I hope some of you will find this patch useful. I've found this discussion page a couple of times in a search engine.

Рус: У меня есть успехи в этом направлении. Я портировал Offscreen драйвер из SDL, но он работает только с EGL. Чтобы включить полноценную поддержку EGL, я обновил версию GLEW и поставил параметр EGL в CMake, но из-за этого он теперь работает только под Wayland и в режиме Offscreen, в Х11 ошибки шейдеров. Но оно неплохо работает в моём ПО для витуберства в сочетании с [obs-vkcapture](https://github.com/nowrep/obs-vkcapture/). Надеюсь кому-нибудь из вас пригодится, эта страница мне выпадала пару раз в поисковике.

https://github.com/scaledteam/urho3d/tree/offscreen-sdl-egl

**Comment by 1vanK** [comment]
Примеры, которые с движом поставляются, нормально собираются?

**Comment by scaledteam** [comment]
Не собираются, даже Urho3D Player не собирается. Вот с такими изменениями собирается:
```
cmake . -DURHO3D_SAMPLES=NO -DURHO3D_PLAYER=NO  -DURHO3D_ANGELSCRIPT=NO -DURHO3D_TOOLS=NO
```

Но вообще я отключаю почти всё когда собираю, так что да, упустил этот момент с линковкой. Хотелось бы как-то по-правильному EGL слинковать.
```
cmake . -DURHO3D_TOOLS=NO -DURHO3D_ANGELSCRIPT=NO -DURHO3D_LUA=NO -DURHO3D_NETWORK=NO -DURHO3D_SAMPLES=NO -DURHO3D_WEBP=NO -DURHO3D_IK=NO -DURHO3D_THREADING=NO -DURHO3D_NAVIGATION=NO -DURHO3D_URHO2D=NO -DURHO3D_LUAJIT=NO -DURHO3D_LUAJIT_AMALG=NO
```

**Comment by 1vanK** [comment]
Уточню, чтобы мы не говорили о разном. Примеры движка без твоих модификаций (из репозитория) не собираются без дополнительных манипуляций?

**Comment by scaledteam** [comment]
Urho3D без изменений собирается отлично. С модификациями не собирается.

**Comment by 1vanK** [comment]
Просто у нас GitHub workflow собирает движок с примерами на виртуальной машине линукс и запускает их после компиляции

**Comment by 1vanK** [comment]
Понял, т.е. вопрос в том, как зависимости пробросить в примеры

**Comment by scaledteam** [comment]
Да, чтобы EGL не надо было в каждом примере указывать, а можно было глобально указать где-нибудь здесь:

https://github.com/scaledteam/urho3d/blob/offscreen-sdl-egl/Source/ThirdParty/GLEW/CMakeLists.txt

**Comment by 1vanK** [comment]
Навскидку не скажу, в UrhoCommon.cmake линкуются либы и к движку и к примерам. Там списки всякие библиотек создаются

**Comment by 1vanK** [comment]
Вот это выглядит как то что нужно

```
# *** THIS IS A DEPRECATED MACRO ***
# Macro for defining external library dependencies
# The purpose of this macro is emulate CMake to set the external library dependencies transitively
# It works for both targets setup within Urho3D project and downstream projects that uses Urho3D as external static/shared library
# *** THIS IS A DEPRECATED MACRO ***
macro (define_dependency_libs TARGET)
```

Написано deprecated, но вроде она везде вызывается

**Comment by scaledteam** [comment]
Я добавил EGL здесь и оно начало собираться. Сейчас ещё свой проект потестирую чтобы убедится что ничего не упустил.
https://github.com/scaledteam/urho3d/blob/offscreen-sdl-egl/cmake/Modules/UrhoCommon.cmake#L968

**Comment by scaledteam** [comment]
Работает однако, спасибо большое!
По-хорошему надо будет это опцией реализовать, потому что EGL вроде как не поддерживается на очень старых системах или где-то ещё, не помню. Под Х11 он поддерживается, но в текущем состоянии выдаёт какие-то странные ошибки:

```
[Fri Jan  6 16:47:50 2023] ERROR: Could not initialize OpenGL extensions, root cause: 'Missing GL version'
[Fri Jan  6 16:47:50 2023] ERROR: Failed to create 2D texture err=1281, target=3553, format=6409, externalFormat=6409, dataType=5121
[Fri Jan  6 16:47:50 2023] ERROR: Failed to create 2D texture err=1281, target=3553, format=6409, externalFormat=6409, dataType=5121
[Fri Jan  6 16:47:50 2023] ERROR: Failed to compile pixel shader Unlit(DIFFMAP):
0:247(16): error: syntax error, unexpected BASIC_TYPE_TOK, expecting '{'
[Fri Jan  6 16:47:50 2023] ERROR: Failed to compile pixel shader Unlit(ALPHAMASK DIFFMAP):
0:248(16): error: syntax error, unexpected BASIC_TYPE_TOK, expecting '{'
[Fri Jan  6 16:47:50 2023] ERROR: Failed to compile pixel shader Unlit():
0:246(16): error: syntax error, unexpected BASIC_TYPE_TOK, expecting '{'
Resolution is 640x480.
```

При этом под Wayland и в режиме Offscreen всё ок. Не подскажешь кстати, с чем может быть связано это?

**Comment by 1vanK** [comment]
Понятия не имею, мб там OpenGL контекст дял EGL как-то иначе создавать надо. И вообще у нас SDL старый щас, обновлять надо

**Comment by 1vanK** [comment]
Я сейчас workflow для обновления года в копирайте ковыряю, закончу, будет время в этом поразбираться

**Comment by scaledteam** [comment]
Да, он староват. Но обновлять его конечно непростая задача. ПО этой причине я портировал [именно драйвер offscreen](https://github.com/libsdl-org/SDL/commit/68985371a071b97b961df56d040fbb5ca6493cf3), а не заменил SDL на более новый, тупо не потянул. Там много изменений Urho3D'шных в CMake, и добавлены дополнительные функции в сам SDL.
